aten/src/ATen/native/cpu/CatKernel.cpp

```diff
@@ -0,0 +1,75 @@
+#include <ATen/ATen.h>
+
+#include <ATen/Dispatch.h>
+#include <ATen/native/cpu/CatKernel.h>
+#include <ATen/cpu/vec256/functional.h>
+#include <ATen/cpu/vec256/vec256.h>
+
+namespace at { namespace native {
+
+namespace {
+
+struct InputMeta {
+  void* data_ptr;
+  int64_t inner_size;
+
+  InputMeta(const Tensor& t, int64_t dim, int64_t inner)
+    : data_ptr(t.data_ptr())
+    , inner_size(t.size(dim) * inner) {}
+};
+
+template <typename scalar_t>
+void cat_serial_kernel_impl(Tensor& result, TensorList tensors, int64_t dim) {
+  auto size = result.sizes().vec();
+  int64_t outer = 1, inner = 1;
+  for (int64_t i = 0; i < dim; i++) {
+    outer *= size[i];
+  }
+  for (int64_t i = dim + 1; i < size.size(); i++) {
+    inner *= size[i];
+  }
+  scalar_t* result_data = result.data_ptr<scalar_t>();
+  int64_t ninputs = tensors.size();
+  std::vector<InputMeta> inputs;
+  inputs.reserve(ninputs);
+  for (auto const &tensor : tensors) {
+    inputs.emplace_back(tensor, dim, inner);
+  }
+ 
+  using Vec = vec256::Vec256<scalar_t>;
+  int64_t offset = 0;
+  for (int64_t i = 0; i < outer; i++) {
+    for (int64_t j = 0; j < ninputs; j++) {
+      scalar_t* result_ptr = result_data + offset;
+      int64_t local_inner = inputs[j].inner_size;
+      scalar_t* input_ptr = (scalar_t*)(inputs[j].data_ptr) + i * local_inner;
+      if (local_inner < Vec::size()) {
+        #ifndef _MSC_VER
+        # pragma unroll
+        #endif
+        for (int64_t k = 0; k < local_inner; k++) {
+          result_ptr[k] = input_ptr[k];
+        }
+      } else {
+        vec256::map(
```

Comment:
@mingfeima does this produce avx move instructions? Can you comment on why memcpy wasn't used here instead? I couldn't see the reasoning in the comments.

Reply:
`std::memcpy` is upto SSE (128b). And vec256::map will compile AVX and AVX2 assembly (pytorch choose AVX or AVX2 according to your cpu ability at runtime). Unfortunately, neither would generate AVX512 code... Anyway, for memory bandwidth bond cases, should not make too much difference.